### PR TITLE
[BEAM-6165] send metrics to Flink in portable Flink runner

### PR DESCRIPTION
**Pull request by ryan-williams**
The Python SDK plumbs user metrics back in `ProgressBundle{Progress,}Response`, but the portable Flink runner doesn't send them on to Flink when executing stages.

This fixes that, and adds a test (in the python SDK) that increments a `Counter` in a `DoFn` and verifies that the result is written to a file by a Flink metrics reporter (`FileReporter`).

R: @angoenka 
CC: @robertwb @tweise @mxm who I've discussed this with
CC: @ajamato who is doing related metrics work (cf. #6799)

Post-Commit Tests Status (on master branch)
------------------------------------------------------------------------------------------------

Lang | SDK | Apex | Dataflow | Flink | Gearpump | Samza | Spark
--- | --- | --- | --- | --- | --- | --- | ---
Go | [Build Status](https://builds.apache.org/job/beam_PostCommit_Go_GradleBuild/lastCompletedBuild/) | --- | --- | --- | --- | --- | ---
Java | [Build Status](https://builds.apache.org/job/beam_PostCommit_Java_GradleBuild/lastCompletedBuild/) | [Build Status](https://builds.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Apex_Gradle/lastCompletedBuild/) | [Build Status](https://builds.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Dataflow_Gradle/lastCompletedBuild/) | [Build Status](https://builds.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Flink_Gradle/lastCompletedBuild/) [Build Status](https://builds.apache.org/job/beam_PostCommit_Java_PVR_Flink/lastCompletedBuild/) | [Build Status](https://builds.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Gearpump_Gradle/lastCompletedBuild/) | [Build Status](https://builds.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Samza_Gradle/lastCompletedBuild/) | [Build Status](https://builds.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Spark_Gradle/lastCompletedBuild/)
Python | [Build Status](https://builds.apache.org/job/beam_PostCommit_Python_Verify/lastCompletedBuild/) | --- | [Build Status](https://builds.apache.org/job/beam_PostCommit_Py_VR_Dataflow/lastCompletedBuild/) </br> [Build Status](https://builds.apache.org/job/beam_PostCommit_Py_ValCont/lastCompletedBuild/) | [Build Status](https://builds.apache.org/job/beam_PostCommit_Python_VR_Flink/lastCompletedBuild/) | --- | --- | ---

### CODE REVIEW

File: runners/flink/src/main/java/org/apache/beam/runners/flink/FlinkExecutionEnvironments.java [hunk 1]
```diff
@@ -127,22 +129,24 @@ static StreamExecutionEnvironment createStreamExecutionEnvironment(
     LOG.info("Creating a Streaming Environment.");
 
     String masterUrl = options.getFlinkMaster();
-    Configuration flinkConfig = getFlinkConfiguration(confDir);
+    Configuration flinkConfiguration = getFlinkConfiguration(confDir);
```
Comment (ryan-williams) [inline runners/flink/src/main/java/org/apache/beam/runners/flink/FlinkExecutionEnvironments.java]: this change is for consistency with the batch environment creation function (also in this file)

File: runners/flink/src/main/java/org/apache/beam/runners/flink/FlinkRunner.java [hunk 2]
```diff
@@ -126,7 +133,43 @@ static PipelineResult createPipelineResult(JobExecutionResult result, PipelineOp
       if (accumulators != null && !accumulators.isEmpty()) {
         LOG.info("Final accumulator values:");
         for (Map.Entry<String, Object> entry : result.getAllAccumulatorResults().entrySet()) {
-          LOG.info("{} : {}", entry.getKey(), entry.getValue());
+          MetricsContainerStepMap stepMap = (MetricsContainerStepMap) entry.getValue();
```
Comment (ryan-williams) [inline runners/flink/src/main/java/org/apache/beam/runners/flink/FlinkRunner.java]: I believe `MetricsContainerStepMap` was just being output as an object address here, so I expanded it to something more meaningful.

This could get factored out to a helper (overriding `MetricsContainerStepMap.toString` to do this feels like too much)

File: runners/flink/src/main/java/org/apache/beam/runners/flink/metrics/FlinkMetricContainer.java [hunk 3]
```diff
@@ -77,12 +82,72 @@ public FlinkMetricContainer(RuntimeContext runtimeContext) {
...
+            String urn = monitoringInfo.getUrn();
+            MetricName metricName = parseUrn(urn);
+            BeamFnApi.Metric metric = monitoringInfo.getMetric();
+            if (metric.hasCounterData()) {
+              BeamFnApi.CounterData counterData = metric.getCounterData();
+              org.apache.beam.sdk.metrics.Counter counter = metricsContainer.getCounter(metricName);
+              if (counterData.getValueCase() == BeamFnApi.CounterData.ValueCase.INT64_VALUE) {
+                counter.inc(counterData.getInt64Value());
+              } else {
+                throw new IllegalArgumentException("Unsupported CounterData type: " + counterData);
```
Comment (ryan-williams) [inline runners/flink/src/main/java/org/apache/beam/runners/flink/metrics/FlinkMetricContainer.java]: `org.apache.beam.sdk.metrics.Counter` only supports `long`s…

File: runners/flink/src/main/java/org/apache/beam/runners/flink/metrics/FlinkMetricContainer.java [hunk 4]
```diff
@@ -77,12 +82,72 @@ public FlinkMetricContainer(RuntimeContext runtimeContext) {
...
+              if (distributionData.hasIntDistributionData()) {
+                BeamFnApi.IntDistributionData intDistributionData =
+                    distributionData.getIntDistributionData();
+                distribution.update(
+                    intDistributionData.getSum(),
+                    intDistributionData.getCount(),
+                    intDistributionData.getMin(),
+                    intDistributionData.getMax());
+              } else {
+                throw new IllegalArgumentException(
```
Comment (ryan-williams) [inline runners/flink/src/main/java/org/apache/beam/runners/flink/metrics/FlinkMetricContainer.java]: likewise, `Distribution` only supports `long`s

File: sdks/python/apache_beam/runners/portability/flink_runner_test.py [hunk 5]
```diff
@@ -59,25 +65,63 @@
...
+      # will be cleaned up at the end of this method, and recreated and used by the job server
+      tmp_dir = mkdtemp(prefix='flinktest')
+
+      cls._create_conf_dir()
+
       try:
         return [
             'java',
             '-jar', flink_job_server_jar,
+            '--flink-master-url', '[local]',
```
Comment (ryan-williams) [inline sdks/python/apache_beam/runners/portability/flink_runner_test.py]: `test_metrics` below only works in `[local]` mode; the Flink `Configuration` [isn't propagated through the `[auto]` mode](https://github.com/apache/beam/pull/7183/files#diff-6acb0479d563cfc121ac04e789f4bc6dR141), but maybe I should make that happen? or is it fine to only run this suite on local Flink clusters?

Comment (tweise) [reply sdks/python/apache_beam/runners/portability/flink_runner_test.py]: This suite already assumes Flink embedded in the job server. If it would run on a remote Flink cluster, a metric reporter that writes to local file would also not work.

File: runners/flink/src/main/java/org/apache/beam/runners/flink/FlinkExecutionEnvironments.java (hunk 1 above)
Comment (tweise) [reply runners/flink/src/main/java/org/apache/beam/runners/flink/FlinkExecutionEnvironments.java]: I don't think this change is necessary for the scope of this PR.

File: sdks/python/build.gradle [hunk 6]
```diff
@@ -407,14 +407,19 @@ def flinkCompatibilityMatrix = {
   }
 }
 
+def flinkCompatibilityMatrixStreamingDOCKER = flinkCompatibilityMatrix(streaming: true)
```
Comment (tweise) [inline sdks/python/build.gradle]: Why this change?

**Comment by tweise** [comment]
@ryan-williams thanks for working on this. I will take a look in the next few days. I appreciate the detailed attention to peripheral code but want to point out that PRs changes should be limited to the stated purpose. Stray, unrelated changes cause unnecessary noise/distraction and can even lead to merge conflicts. It would be nice to make such changes in a separate PR. IMO there is no need to back them out in this case unless specifically noted, but please keep it in mind for future work. Thanks!

File: sdks/python/apache_beam/runners/portability/flink_runner_test.py (hunk 5 above)
Comment (ryan-williams) [reply sdks/python/apache_beam/runners/portability/flink_runner_test.py]: I think we are agreeing that this change is OK: this suite is only run (or intended to run) on local Flink clusters. 

I hadn't seen any extant part of it that would fail on a remote cluster, but you are saying such a part exists, IIUC.

I agree that my metric reporter would not work "as is" if this suite was run on a remote Flink cluster.

File: runners/flink/src/main/java/org/apache/beam/runners/flink/metrics/FlinkMetricContainer.java [hunk 7]
```diff
@@ -77,12 +82,72 @@ public FlinkMetricContainer(RuntimeContext runtimeContext) {
...
         : null;
   }
 
+  /**
+   * Parse a {@link MetricName} from a {@link
+   * org.apache.beam.model.fnexecution.v1.BeamFnApi.MonitoringInfoUrns.Enum}
+   *
+   * <p>Should be consistent with {@code parse_namespace_and_name} in monitoring_infos.py
+   *
+   * <p>TODO: not flink-specific; where should it live?
```
Comment (ryan-williams) [inline runners/flink/src/main/java/org/apache/beam/runners/flink/metrics/FlinkMetricContainer.java]: Flagging this as well, in case anyone has ideas.

I would normally view this as a`MetricName` constructor, but `MetricName` is in java-core, which doesn't depend on fn-api (and shouldn't be made to without good reason, I assume).

**Comment by ryan-williams** [comment]
makes sense @tweise, and thanks for the quick look. I went ahead and removed all but the essential changes.

previously there was [a `beam_PreCommit_JavaPortabilityApi` failure](https://builds.apache.org/job/beam_PreCommit_JavaPortabilityApi_Commit/545/console) that I'm hoping was unrelated flakiness.

File: runners/flink/src/main/java/org/apache/beam/runners/flink/metrics/FlinkMetricContainer.java (hunk 7 above)
Comment (mxm) [reply runners/flink/src/main/java/org/apache/beam/runners/flink/metrics/FlinkMetricContainer.java]: I believe that should be changed in `BeamFnApi.MonitoringInfo`.  Ok for now as a workaround.

File: runners/flink/src/main/java/org/apache/beam/runners/flink/FlinkExecutionEnvironments.java [hunk 8]
```diff
@@ -132,7 +134,9 @@ static StreamExecutionEnvironment createStreamExecutionEnvironment(
 
     // depending on the master, create the right environment.
     if ("[local]".equals(masterUrl)) {
-      flinkStreamEnv = StreamExecutionEnvironment.createLocalEnvironment();
+      flinkStreamEnv =
+          StreamExecutionEnvironment.createLocalEnvironment(
+              getDefaultLocalParallelism(), flinkConfig);
```
[the remark this answers is not in the document]
Comment (tweise) [reply runners/flink/src/main/java/org/apache/beam/runners/flink/FlinkExecutionEnvironments.java]: I think this can be avoided by setting the `FLINK_CONF_DIR` when launching the job server.

File: sdks/python/apache_beam/runners/portability/flink_runner_test.py [hunk 9]
```diff
@@ -65,19 +69,58 @@ class FlinkRunnerTest(portable_runner_test.PortableRunnerTest):
...
+    @classmethod
+    def tearDownClass(cls):
+      if cls.conf_dir and exists(cls.conf_dir):
+        logging.info("removing conf dir: %s" % cls.conf_dir)
+        rmtree(cls.conf_dir)
+      super(FlinkRunnerTest, cls).tearDownClass()
+
+    @classmethod
+    def _create_conf_dir(cls):
+      """Create (and save a static reference to) a "conf dir", used to provide
```
Comment (tweise) [inline sdks/python/apache_beam/runners/portability/flink_runner_test.py]: This looks convoluted. Could we just check in the file and point to that directory instead?

Comment (ryan-williams) [reply sdks/python/apache_beam/runners/portability/flink_runner_test.py]: good idea, though I am also using this directory as the tempdir that holds the output metrics; I think I'd still need to do that and it would look the same as this, but perhaps with a different name.

File: runners/flink/src/main/java/org/apache/beam/runners/flink/metrics/FileReporter.java [hunk 10]
```diff
@@ -0,0 +1,74 @@
...
+import org.apache.flink.metrics.Metric;
+import org.apache.flink.metrics.MetricConfig;
+import org.apache.flink.metrics.MetricGroup;
+import org.apache.flink.metrics.reporter.AbstractReporter;
+
+/**
+ * Flink {@link org.apache.flink.metrics.reporter.MetricReporter metrics reporter} for writing
+ * metrics to a file (specified via the "metrics.reporter.test.file" config key).
+ */
+public class FileReporter extends AbstractReporter {
```
[the remark this answers is not in the document]
Comment (ryan-williams) [reply runners/flink/src/main/java/org/apache/beam/runners/flink/metrics/FileReporter.java]: I originally gave it a name like that, however I talked myself in to that this is a potentially more general-purpose abstraction, and it is Flink specific, so that this (`main`, not `test`) package is an OK place for it.

If it was `TestFileReporter` then I'd feel obliged to put it in the test package, at which point the build plumbing to start a job-server in `flink_runner_test.py` that has the :beam-runners-flink` `-tests.jar` on the classpath would become more convoluted, so that was an added incentive to view this as not a "tests only" thing 😀

lmk what you think!

File: runners/flink/src/main/java/org/apache/beam/runners/flink/metrics/FlinkMetricContainer.java [hunk 11]
```diff
@@ -77,12 +82,72 @@ public FlinkMetricContainer(RuntimeContext runtimeContext) {
...
     return metricsAccumulator != null
         ? metricsAccumulator.getLocalValue().getContainer(stepName)
         : null;
   }
 
+  /**
+   * Parse a {@link MetricName} from a {@link
+   * org.apache.beam.model.fnexecution.v1.BeamFnApi.MonitoringInfoUrns.Enum}
+   *
+   * <p>Should be consistent with {@code parse_namespace_and_name} in monitoring_infos.py
```
Comment (ryan-williams) [inline runners/flink/src/main/java/org/apache/beam/runners/flink/metrics/FlinkMetricContainer.java]: Wanted to flag this for @ajamato in particular as the author of [the python code I mimicked here](https://github.com/apache/beam/blob/6b7cf422733a82aabb8e257e975d9e9c5e785376/sdks/python/apache_beam/metrics/monitoring_infos.py#L228-L236).

How important is it that the python and java SDKs parse metric URNs into a {namespace, name} pairs in the same way?

From what I can tell, It may not be critical, though we should still probably agree on one semantic interpretation of what is "namespace" vs "name" of e.g. [`"beam:metric:element_count:v1"`](https://github.com/apache/beam/blob/master/model/fn-execution/src/main/proto/beam_fn_api.proto#L360-L361).

With `monitoring_infos.py` as the reference implementation, the answer is {`"beam"`, `"metric:element_count:v1"`}; shall I encode that in a comment in `beam_fn_api.proto`, and reference it here?

File: runners/flink/src/main/java/org/apache/beam/runners/flink/FlinkExecutionEnvironments.java (hunk 8 above)
[the remark this answers is not in the document]
Comment (ryan-williams) [reply runners/flink/src/main/java/org/apache/beam/runners/flink/FlinkExecutionEnvironments.java]: @mxm previously, a Flink `Configuration` in scope was thrown away in favor of a `new Configuration()`, dropping settings from `--flink-conf-dir` that were [supposed to override `$FLINK_CONF_DIR`](https://github.com/apache/beam/pull/7031#issuecomment-442112065), and generally going against the patterns I know for building up a config object like this from sources with different precedence.

So this seemed like a bug-fix worth doing in general, which also blocked my metrics configs from propagating as needed here.

@tweise `$FLINK_CONF_DIR` vs `--flink-conf-dir` is an interesting question… is one clearly preferred? I thought the latter was more explicit, easier to pass to the job-server, and overrides the former, but I didn't fully follow the rationale around it in #7031. 

Also if I factored the metrics YAML file out as a test resource [like you suggested below](https://github.com/apache/beam/pull/7183/files#r238293952), I don't see it being any easier to point at via `$FLINK_CONF_DIR` vs `--flink-conf-dir`, but I may be missing things!

lmk what you think, thanks

File: sdks/python/apache_beam/runners/portability/flink_runner_test.py (hunk 5 above)
Comment (mxm) [reply sdks/python/apache_beam/runners/portability/flink_runner_test.py]: Ah, this is the one I was referring to: #7031

File: runners/flink/src/main/java/org/apache/beam/runners/flink/metrics/FileReporter.java (hunk 10 above)
[the remark this answers is not in the document]
Comment (mxm) [reply runners/flink/src/main/java/org/apache/beam/runners/flink/metrics/FileReporter.java]: I get why it's not practical to have it in the test scope. That's fine. Still, I think `TestFileReporter` reflects more accurately what it does. Also, it's configured via a test property, so I don't see how it is useful for non-testing.

File: runners/flink/src/main/java/org/apache/beam/runners/flink/FlinkExecutionEnvironments.java (hunk 8 above)
[the remark this answers is not in the document]
Comment (mxm) [reply runners/flink/src/main/java/org/apache/beam/runners/flink/FlinkExecutionEnvironments.java]: Makes sense. Thanks for clarifying.

I'd say `--flink-conf-dir` is now the preferred approach.

File: runners/flink/src/main/java/org/apache/beam/runners/flink/metrics/FileReporter.java (hunk 10 above)
[the remark this answers is not in the document]
Comment (ryan-williams) [reply runners/flink/src/main/java/org/apache/beam/runners/flink/metrics/FileReporter.java]: > it's configured via a test property

ah, [I left the reporter alias as `test` in `flink_runner_test`](https://github.com/apache/beam/pull/7183/files#diff-195e4a1918aec54f6245e49081cadb5aR99) when I renamed this class to remove the `Test` prefix. I should change that alias to `file`, either way.

This is a reporter that writes the final state of all metrics to a file of the user's choosing. It can do this in test or production environments. It doesn't perform any assertions, or otherwise do things that only make sense in a testing context. It is fairly similar to [Slf4jReporter](https://github.com/apache/flink/blob/release-1.5.5/flink-metrics/flink-metrics-slf4j/src/main/java/org/apache/flink/metrics/slf4j/Slf4jReporter.java#L41), but its output is easier to parse (for machines or humans), and would normally be configured like:

```
metrics.reporters: file,
metrics.reporter.file.class: org.apache.beam.runners.flink.metrics.FileReporter,
metrics.reporter.file.file: /some/path/on/job-mgr/node
```

To me, it feels weirder to name it `Test*` and leave it under `main/`, when it is not really *that* test-specific an abstraction.

However I understand if the possible non-test uses feel contrived, and perhaps having `Test*` names under `main/` isn't too smelly.

I just want to make sure the vestigial `metrics.reporter.test.*` namespacing isn't driving the decision too heavily.

File: sdks/python/apache_beam/runners/portability/flink_runner_test.py (hunk 9 above)
Comment (tweise) [reply sdks/python/apache_beam/runners/portability/flink_runner_test.py]: makes sense

File: runners/flink/src/main/java/org/apache/beam/runners/flink/metrics/FileReporter.java (hunk 10 above)
[the remark this answers is not in the document]
Comment (tweise) [reply runners/flink/src/main/java/org/apache/beam/runners/flink/metrics/FileReporter.java]: That sounds reasonable. But shouldn't we use a more distinct name to avoid confusion, like `BeamFlinkFileMetricReporter`?

**Comment by ryan-williams** [comment]
Run Java_Examples_Dataflow PreCommit

File: runners/flink/src/main/java/org/apache/beam/runners/flink/metrics/FileReporter.java (hunk 10 above)
[the remark this answers is not in the document]
Comment (mxm) [reply runners/flink/src/main/java/org/apache/beam/runners/flink/metrics/FileReporter.java]: I suppose it could come in handy when you just want to dump metrics locally.

Since package is `org.apache.beam.runners.flink.metrics`, `FileReporter` seems fine.

[the remark this answers is not in the document]
Comment (ryan-williams) [reply runners/flink/src/main/java/org/apache/beam/runners/flink/metrics/FileReporter.java]: fwiw: "beam", "flink", "metrics", "file", and "reporter" are already in this class' name: `org.apache.beam.runners.flink.metrics.FileReporter` (and users would use this FQN in `flink-conf.yaml`, as in the example above).

I generally dislike the Java convention of repeating package-name segments in class' basenames, but recognize that it's common and people seem to like it, so happy to add here, just wanted to float the idea that `org.apache.beam.runners.flink.metrics.BeamFlinkFileMetricReporter` is redundant.

lmk if you still prefer it and I'll change it.

File: sdks/python/apache_beam/runners/portability/flink_runner_test.py (hunk 5 above)
Comment (ryan-williams) [reply sdks/python/apache_beam/runners/portability/flink_runner_test.py]: thanks, I am using the `--flink-conf-dir` added in #7031; I don't see a way to avoid specifying `[local]` here though. it also sounds like we are saying that this test is intended to only run in local mode, so we may as well make that explicit?

The previous (`[auto]`) code path goes through Flink code that doesn't seem to support injecting an existing `Configuration` (cf. [streaming](https://github.com/apache/flink/blob/release-1.5.5/flink-streaming-java/src/main/java/org/apache/flink/streaming/api/environment/StreamExecutionEnvironment.java#L1595-L1612), [batch](https://github.com/apache/flink/blob/release-1.5.5/flink-java/src/main/java/org/apache/flink/api/java/ExecutionEnvironment.java#L1058-L1061); presumably this makes sense because it would be strange to do that if actually in remote/cluster mode?)

lmk if that doesn't make sense or there is something I should change here, thanks

**Comment by tweise** [comment]
Run Python Flink ValidatesRunner

File: runners/flink/src/main/java/org/apache/beam/runners/flink/metrics/FileReporter.java (hunk 10 above)
[the remark this answers is not in the document]
Comment (tweise) [reply runners/flink/src/main/java/org/apache/beam/runners/flink/metrics/FileReporter.java]: Reason I brought it up is discoverability. Names like `Node`, `Metric`, `Reporter` all usually too generic, regardless in which package they reside. When you work in the IDE, you would probably not want to see dozens of matches when searching a class? I checked and in this case it is not the case however and not critical for this PR anyways. BTW it seems to be a convention in Beam code to prefix classes with the runner (like `Flink` when they are likely to appear in multiple runners).

**Comment by tweise** [comment]
@ryan-williams thanks for this contribution!

File: runners/flink/src/main/java/org/apache/beam/runners/flink/metrics/FlinkMetricContainer.java [hunk 12]
```diff
@@ -77,12 +82,72 @@ public FlinkMetricContainer(RuntimeContext runtimeContext) {
...
+   *
+   * <p>TODO: not flink-specific; where should it live?
+   */
+  public static MetricName parseUrn(String urn) {
+    if (urn.startsWith(USER_COUNTER_URN_PREFIX.toString())) {
```
Comment (ajamato) [inline runners/flink/src/main/java/org/apache/beam/runners/flink/metrics/FlinkMetricContainer.java]: Does calling toString on this work? The usage I have seen before involved extracting the attached object and pulling out a field from the enum.

See BeamUrns.getUrns
https://github.com/apache/beam/blob/a5a139d5c7be01fbbd1f8b14df9ef5e05696af2c/runners/core-construction-java/src/main/java/org/apache/beam/runners/core/construction/BeamUrns.java#L27

You could reference this file which pulls it in. SimpleMonitoringInfoBuilder

https://github.com/apache/beam/blob/dc10f75a7bda91f251d43309a21327fc2ab26154/runners/core-java/src/main/java/org/apache/beam/runners/core/metrics/SimpleMonitoringInfoBuilder.java#L64

Comment (ryan-williams) [reply runners/flink/src/main/java/org/apache/beam/runners/flink/metrics/FlinkMetricContainer.java]: good catch, thanks! I just filed #7408 to fix
